File: backend/src/services/ocr.service.ts
/**
 * OCR Service using AWS Textract
 * Extracts text from documents with confidence scoring
 */

import { TextractClient, DetectDocumentTextCommand, AnalyzeDocumentCommand } from '@aws-sdk/client-textract';
import { createLogger } from '../utils/logger';

const logger = createLogger('ocr-service');

export interface OCRResult {
  text: string;
  confidence: number;
  pages: number;
  extractedData?: {
    keyValuePairs?: Array<{ key: string; value: string; confidence: number }>;
    tables?: Array<{ rows: number; columns: number; data: string[][] }>;
  };
  processingTime: number;
}

export class OCRService {
  private textractClient: TextractClient;
  private ocrEnabled: boolean;

  constructor() {
    this.ocrEnabled = process.env.OCR_ENABLED !== 'false';

    if (this.ocrEnabled) {
      this.textractClient = new TextractClient({
        region: process.env.AWS_REGION || 'us-east-1',
        credentials: {
          accessKeyId: process.env.AWS_ACCESS_KEY_ID || '',
          secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || '',
        },
      });

      logger.info('AWS Textract OCR service initialized');
    } else {
      logger.warn('OCR disabled - development mode');
    }
  }

  /**
   * Extract text from document using AWS Textract
   */
  async extractText(fileBuffer: Buffer, filename: string): Promise<OCRResult> {
    const startTime = Date.now();

    try {
      if (!this.ocrEnabled) {
        logger.info(`OCR bypassed for: ${filename} (development mode)`);
        return {
          text: '',
          confidence: 0,
          pages: 0,
          processingTime: Date.now() - startTime,
        };
      }

      // Determine document type
      const isImage = this.isImageFile(filename);

      // Use appropriate Textract API
      const result = isImage
        ? await this.detectDocumentText(fileBuffer)
        : await this.analyzeDocument(fileBuffer);

      const processingTime = Date.now() - startTime;

      logger.info(`OCR completed for ${filename}: ${result.text.length} chars (${processingTime}ms)`);

      return {
        ...result,
        processingTime,
      };
    } catch (error) {
      logger.error(`OCR error for ${filename}:`, error);
      throw new Error(`OCR processing failed: ${error}`);
    }
  }

  /**
   * Simple text detection for images
   */
  private async detectDocumentText(fileBuffer: Buffer): Promise<OCRResult> {
    const command = new DetectDocumentTextCommand({
      Document: {
        Bytes: fileBuffer,
      },
    });

    const response = await this.textractClient.send(command);

    // Extract text and calculate average confidence
    const blocks = response.Blocks || [];
    const textBlocks = blocks.filter((block) => block.BlockType === 'LINE');

    const text = textBlocks
      .map((block) => block.Text)
      .filter(Boolean)
      .join('\n');

    const confidences = textBlocks
      .map((block) => block.Confidence || 0)
      .filter((c) => c > 0);

    const avgConfidence =
      confidences.length > 0
        ? confidences.reduce((a, b) => a + b, 0) / confidences.length
        : 0;

    return {
      text,
      confidence: avgConfidence / 100, // Convert to 0-1 scale
      pages: 1,
      processingTime: 0, // Will be set by caller
    };
  }

  /**
   * Advanced document analysis for PDFs and complex documents
   */
  private async analyzeDocument(fileBuffer: Buffer): Promise<OCRResult> {
    const command = new AnalyzeDocumentCommand({
      Document: {
        Bytes: fileBuffer,
      },
      FeatureTypes: ['FORMS', 'TABLES'], // Extract forms and tables
    });

    const response = await this.textractClient.send(command);

    const blocks = response.Blocks || [];

    // Extract text
    const textBlocks = blocks.filter((block) => block.BlockType === 'LINE');
    const text = textBlocks
      .map((block) => block.Text)
      .filter(Boolean)
      .join('\n');

    // Calculate confidence
    const confidences = textBlocks
      .map((block) => block.Confidence || 0)
      .filter((c) => c > 0);

    const avgConfidence =
      confidences.length > 0
        ? confidences.reduce((a, b) => a + b, 0) / confidences.length
        : 0;

    // Extract key-value pairs (forms)
    const keyValuePairs = this.extractKeyValuePairs(blocks);

    // Extract tables
    const tables = this.extractTables(blocks);

    return {
      text,
      confidence: avgConfidence / 100,
      pages: response.DocumentMetadata?.Pages || 1,
      extractedData: {
        keyValuePairs,
        tables,
      },
      processingTime: 0,
    };
  }

  /**
   * Extract key-value pairs from form data
   */
  private extractKeyValuePairs(
    blocks: any[]
  ): Array<{ key: string; value: string; confidence: number }> {
    const kvPairs: Array<{ key: string; value: string; confidence: number }> = [];

    const keyBlocks = blocks.filter(
      (block) => block.BlockType === 'KEY_VALUE_SET' && block.EntityTypes?.includes('KEY')
    );

    for (const keyBlock of keyBlocks) {
      if (!keyBlock.Relationships) continue;

      // Find the associated value block
      const valueRelation = keyBlock.Relationships.find(
        (rel: any) => rel.Type === 'VALUE'
      );

      if (!valueRelation) continue;

      const valueBlockId = valueRelation.Ids?.[0];
      const valueBlock = blocks.find((b) => b.Id === valueBlockId);

      if (!valueBlock) continue;

      // Extract text from child blocks
      const keyText = this.extractChildText(blocks, keyBlock);
      const valueText = this.extractChildText(blocks, valueBlock);

      if (keyText && valueText) {
        kvPairs.push({
          key: keyText,
          value: valueText,
          confidence: Math.min(keyBlock.Confidence || 0, valueBlock.Confidence || 0) / 100,
        });
      }
    }

    return kvPairs;
  }

  /**
   * Extract tables from document
   */
  private extractTables(blocks: any[]): Array<{ rows: number; columns: number; data: string[][] }> {
    const tables: Array<{ rows: number; columns: number; data: string[][] }> = [];

    const tableBlocks = blocks.filter((block) => block.BlockType === 'TABLE');

    for (const tableBlock of tableBlocks) {
      if (!tableBlock.Relationships) continue;

      const cellRelation = tableBlock.Relationships.find((rel: any) => rel.Type === 'CHILD');

      if (!cellRelation) continue;

      const cellIds = cellRelation.Ids || [];
      const cells = blocks.filter((b) => cellIds.includes(b.Id));

      // Build table structure
      const tableData: string[][] = [];
      let maxRow = 0;
      let maxCol = 0;

      for (const cell of cells) {
        if (cell.BlockType !== 'CELL') continue;

        const row = cell.RowIndex || 1;
        const col = cell.ColumnIndex || 1;

        maxRow = Math.max(maxRow, row);
        maxCol = Math.max(maxCol, col);

        if (!tableData[row - 1]) {
          tableData[row - 1] = [];
        }

        tableData[row - 1][col - 1] = this.extractChildText(blocks, cell);
      }

      tables.push({
        rows: maxRow,
        columns: maxCol,
        data: tableData,
      });
    }

    return tables;
  }

  /**
   * Extract text from child blocks
   */
  private extractChildText(blocks: any[], parentBlock: any): string {
    if (!parentBlock.Relationships) return '';

    const childRelation = parentBlock.Relationships.find((rel: any) => rel.Type === 'CHILD');

    if (!childRelation) return '';

    const childIds = childRelation.Ids || [];
    const childBlocks = blocks.filter((b) => childIds.includes(b.Id));

    return childBlocks
      .map((block) => block.Text)
      .filter(Boolean)
      .join(' ');
  }

  /**
   * Check if file is an image
   */
  private isImageFile(filename: string): boolean {
    const imageExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.webp'];
    return imageExtensions.some((ext) => filename.toLowerCase().endsWith(ext));
  }

  /**
   * Check if OCR is applicable for this file type
   */
  isOCRApplicable(mimetype: string): boolean {
    const ocrTypes = [
      'application/pdf',
      'image/jpeg',
      'image/png',
      'image/gif',
      'image/webp',
    ];

    return ocrTypes.includes(mimetype);
  }
}

// Export singleton instance
export const ocrService = new OCRService();

File: ml/src/document_ocr/ocr_service.py
"""
OCR Service with Hybrid Approach
Tesseract for basic documents, AWS Textract for complex documents
Automatic fallback and cost optimization
"""

import pytesseract
from PIL import Image
import boto3
import cv2
import numpy as np
from typing import Dict, List, Optional, Tuple
import logging
import time
import os

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class OCRService:
    """
    Hybrid OCR service combining Tesseract and AWS Textract
    - Tesseract: Fast, free, good for clear text
    - Textract: Advanced, paid, excellent for forms/tables/handwriting
    """

    def __init__(self):
        """Initialize OCR services"""
        self.textract_client = boto3.client('textract', region_name=os.getenv('AWS_REGION', 'us-east-1'))
        self.s3_client = boto3.client('s3', region_name=os.getenv('AWS_REGION', 'us-east-1'))

        # Cost tracking (approximate)
        self.textract_cost_per_page = 0.0015  # $1.50 per 1000 pages

        logger.info('OCR Service initialized with Tesseract and AWS Textract')

    def preprocess_image(self, image_path: str) -> np.ndarray:
        """
        Enhance image quality for better OCR

        Steps:
        1. Grayscale conversion
        2. Noise reduction
        3. Adaptive thresholding
        4. Deskewing

        Args:
            image_path: Path to image file

        Returns:
            Preprocessed image as numpy array
        """
        try:
            img = cv2.imread(image_path)

            if img is None:
                raise ValueError(f"Failed to load image: {image_path}")

            # Convert to grayscale
            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

            # Denoise
            denoised = cv2.fastNlMeansDenoising(gray)

            # Adaptive thresholding
            thresh = cv2.adaptiveThreshold(
                denoised, 255,
                cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
                cv2.THRESH_BINARY, 11, 2
            )

            # Deskew
            angle = self._calculate_skew_angle(thresh)
            rotated = self._rotate_image(thresh, angle)

            logger.info(f"Image preprocessing complete - deskewed by {angle:.2f} degrees")
            return rotated

        except Exception as e:
            logger.error(f"Image preprocessing failed: {str(e)}")
            # Return original grayscale if preprocessing fails
            img = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
            return img

    def _calculate_skew_angle(self, image: np.ndarray) -> float:
        """Calculate skew angle of document"""
        try:
            coords = np.column_stack(np.where(image > 0))
            angle = cv2.minAreaRect(coords)[-1]

            if angle < -45:
                angle = -(90 + angle)
            else:
                angle = -angle

            return angle
        except:
            return 0.0

    def _rotate_image(self, image: np.ndarray, angle: float) -> np.ndarray:
        """Rotate image by given angle"""
        (h, w) = image.shape[:2]
        center = (w // 2, h // 2)
        M = cv2.getRotationMatrix2D(center, angle, 1.0)
        rotated = cv2.warpAffine(
            image, M, (w, h),
            flags=cv2.INTER_CUBIC,
            borderMode=cv2.BORDER_REPLICATE
        )
        return rotated

    def tesseract_ocr(self, image_path: str, preprocess: bool = True) -> Dict:
        """
        Perform OCR using Tesseract

        Args:
            image_path: Path to image file
            preprocess: Whether to preprocess image

        Returns:
            OCR results with text, confidence, and metadata
        """
        start_time = time.time()

        try:
            # Preprocess image if requested
            if preprocess:
                image = self.preprocess_image(image_path)
            else:
                image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)

            # Get detailed OCR data with confidence
            data = pytesseract.image_to_data(
                image,
                output_type=pytesseract.Output.DICT,
                config='--psm 1 --oem 3'  # Auto page segmentation with LSTM engine
            )

            # Extract full text
            full_text = pytesseract.image_to_string(image)

            # Calculate average confidence
            confidences = [int(conf) for conf in data['conf'] if int(conf) > 0]
            avg_confidence = sum(confidences) / len(confidences) if confidences else 0

            # Extract words with positions
            words = []
            for i in range(len(data['text'])):
                if data['text'][i].strip():
                    words.append({
                        'text': data['text'][i],
                        'confidence': data['conf'][i] / 100,
                        'bbox': {
                            'x': data['left'][i],
                            'y': data['top'][i],
                            'width': data['width'][i],
                            'height': data['height'][i]
                        }
                    })

            processing_time = int((time.time() - start_time) * 1000)

            return {
                'provider': 'TESSERACT',
                'full_text': full_text,
                'confidence': avg_confidence / 100,
                'pages': [{
                    'page_number': 1,
                    'text': full_text,
                    'words': words,
                    'confidence': avg_confidence / 100
                }],
                'processing_time': processing_time,
                'cost': 0.0,  # Tesseract is free
                'preprocessing_applied': preprocess
            }

        except Exception as e:
            logger.error(f"Tesseract OCR failed: {str(e)}")
            raise

    def textract_ocr(self, image_path: str, bucket: str, key: str,
                     extract_forms: bool = True, extract_tables: bool = True) -> Dict:
        """
        Perform OCR using AWS Textract

        Args:
            image_path: Path to image file
            bucket: S3 bucket name
            key: S3 object key
            extract_forms: Extract key-value pairs
            extract_tables: Extract tables

        Returns:
            OCR results with text, forms, tables, and metadata
        """
        start_time = time.time()

        try:
            # Upload to S3
            self.s3_client.upload_file(image_path, bucket, key)
            logger.info(f"Uploaded to S3: s3://{bucket}/{key}")

            # Determine feature types
            features = []
            if extract_forms:
                features.append('FORMS')
            if extract_tables:
                features.append('TABLES')

            # Start Textract analysis
            response = self.textract_client.analyze_document(
                Document={'S3Object': {'Bucket': bucket, 'Name': key}},
                FeatureTypes=features
            )

            # Parse results
            result = self._parse_textract_result(response)

            processing_time = int((time.time() - start_time) * 1000)

            # Calculate cost
            page_count = response.get('DocumentMetadata', {}).get('Pages', 1)
            cost = page_count * self.textract_cost_per_page

            result['processing_time'] = processing_time
            result['cost'] = cost

            logger.info(f"Textract OCR complete - {page_count} pages, ${cost:.4f}")

            return result

        except Exception as e:
            logger.error(f"Textract OCR failed: {str(e)}")
            raise

    def _parse_textract_result(self, response: Dict) -> Dict:
        """Parse Textract API response"""
        blocks = response.get('Blocks', [])

        # Extract lines of text
        lines = [block['Text'] for block in blocks if block['BlockType'] == 'LINE']
        full_text = '\n'.join(lines)

        # Extract key-value pairs
        key_values = self._extract_textract_key_values(blocks)

        # Extract tables
        tables = self._extract_textract_tables(blocks)

        # Calculate confidence
        confidences = [block['Confidence'] for block in blocks if 'Confidence' in block]
        avg_confidence = sum(confidences) / len(confidences) if confidences else 0

        # Get page count
        page_count = response.get('DocumentMetadata', {}).get('Pages', 1)

        return {
            'provider': 'AWS_TEXTRACT',
            'full_text': full_text,
            'confidence': avg_confidence / 100,
            'pages': [{
                'page_number': 1,
                'text': full_text,
                'confidence': avg_confidence / 100
            }],
            'key_values': key_values,
            'tables': tables,
            'page_count': page_count
        }

    def _extract_textract_key_values(self, blocks: List[Dict]) -> List[Dict]:
        """Extract key-value pairs from Textract blocks"""
        kv_pairs = []

        key_blocks = [b for b in blocks
                      if b['BlockType'] == 'KEY_VALUE_SET' and 'KEY' in b.get('EntityTypes', [])]

        for key_block in key_blocks:
            if 'Relationships' not in key_block:
                continue

            # Find associated value
            value_relation = next(
                (r for r in key_block['Relationships'] if r['Type'] == 'VALUE'),
                None
            )

            if not value_relation:
                continue

            value_block_id = value_relation['Ids'][0] if value_relation['Ids'] else None
            value_block = next((b for b in blocks if b['Id'] == value_block_id), None)

            if not value_block:
                continue

            # Extract text from child blocks
            key_text = self._extract_child_text(blocks, key_block)
            value_text = self._extract_child_text(blocks, value_block)

            if key_text and value_text:
                kv_pairs.append({
                    'key': key_text,
                    'value': value_text,
                    'confidence': min(
                        key_block.get('Confidence', 0),
                        value_block.get('Confidence', 0)
                    ) / 100
                })

        return kv_pairs

    def _extract_textract_tables(self, blocks: List[Dict]) -> List[Dict]:
        """Extract tables from Textract blocks"""
        tables = []

        table_blocks = [b for b in blocks if b['BlockType'] == 'TABLE']

        for table_block in table_blocks:
            if 'Relationships' not in table_block:
                continue

            cell_relation = next(
                (r for r in table_block['Relationships'] if r['Type'] == 'CHILD'),
                None
            )

            if not cell_relation:
                continue

            cell_ids = cell_relation.get('Ids', [])
            cells = [b for b in blocks if b['Id'] in cell_ids]

            # Build table structure
            table_data = {}
            max_row = 0
            max_col = 0

            for cell in cells:
                if cell['BlockType'] != 'CELL':
                    continue

                row = cell.get('RowIndex', 1)
                col = cell.get('ColumnIndex', 1)

                max_row = max(max_row, row)
                max_col = max(max_col, col)

                if row not in table_data:
                    table_data[row] = {}

                table_data[row][col] = self._extract_child_text(blocks, cell)

            # Convert to array format
            rows = []
            headers = []

            for row_idx in sorted(table_data.keys()):
                row_data = [table_data[row_idx].get(col, '') for col in range(1, max_col + 1)]
                if row_idx == 1:
                    headers = row_data
                else:
                    rows.append(row_data)

            tables.append({
                'row_count': max_row,
                'column_count': max_col,
                'headers': headers,
                'rows': rows,
                'confidence': table_block.get('Confidence', 0) / 100
            })

        return tables

    def _extract_child_text(self, blocks: List[Dict], parent_block: Dict) -> str:
        """Extract text from child blocks"""
        if 'Relationships' not in parent_block:
            return ''

        child_relation = next(
            (r for r in parent_block['Relationships'] if r['Type'] == 'CHILD'),
            None
        )

        if not child_relation:
            return ''

        child_ids = child_relation.get('Ids', [])
        child_blocks = [b for b in blocks if b['Id'] in child_ids]

        return ' '.join(block.get('Text', '') for block in child_blocks).strip()

    def hybrid_ocr(self, image_path: str, bucket: str = None, key: str = None,
                   confidence_threshold: float = 0.85) -> Dict:
        """
        Hybrid OCR approach: Try Tesseract first, fallback to Textract if needed

        Strategy:
        1. Run Tesseract OCR (fast, free)
        2. Check confidence score
        3. If confidence >= threshold, return Tesseract result
        4. Otherwise, use Textract (slower, paid, better quality)

        Args:
            image_path: Path to image file
            bucket: S3 bucket name (required for Textract fallback)
            key: S3 object key (required for Textract fallback)
            confidence_threshold: Minimum confidence to accept Tesseract result

        Returns:
            OCR results with provider information
        """
        logger.info(f"Starting hybrid OCR for {image_path}")

        # Try Tesseract first
        try:
            tesseract_result = self.tesseract_ocr(image_path)

            logger.info(f"Tesseract confidence: {tesseract_result['confidence']:.2f}")

            # If confidence is high enough, use Tesseract
            if tesseract_result['confidence'] >= confidence_threshold:
                logger.info("Using Tesseract result (high confidence)")
                return tesseract_result

            logger.info(f"Tesseract confidence ({tesseract_result['confidence']:.2f}) " +
                       f"below threshold ({confidence_threshold}), falling back to Textract")

        except Exception as e:
            logger.warning(f"Tesseract failed: {str(e)}, falling back to Textract")

        # Fallback to Textract
        if not bucket or not key:
            logger.error("Textract fallback requires bucket and key parameters")
            # Return Tesseract result as fallback
            return tesseract_result if 'tesseract_result' in locals() else {
                'provider': 'HYBRID',
                'full_text': '',
                'confidence': 0.0,
                'error': 'OCR failed'
            }

        try:
            textract_result = self.textract_ocr(image_path, bucket, key)
            textract_result['provider'] = 'HYBRID'
            textract_result['fallback_used'] = True
            textract_result['fallback_reason'] = 'Low Tesseract confidence'

            logger.info("Using Textract result (fallback)")
            return textract_result

        except Exception as e:
            logger.error(f"Textract fallback also failed: {str(e)}")

            # Return best available result
            if 'tesseract_result' in locals():
                tesseract_result['provider'] = 'HYBRID'
                tesseract_result['error'] = 'Textract fallback failed'
                return tesseract_result
            else:
                return {
                    'provider': 'HYBRID',
                    'full_text': '',
                    'confidence': 0.0,
                    'error': 'Both Tesseract and Textract failed'
                }


# Example usage
if __name__ == '__main__':
    ocr = OCRService()

    # Test with sample image
    result = ocr.hybrid_ocr(
        'sample_document.pdf',
        bucket='my-bucket',
        key='documents/sample.pdf'
    )

    print(f"Provider: {result['provider']}")
    print(f"Confidence: {result['confidence']:.2f}")
    print(f"Text length: {len(result['full_text'])} characters")
